File: clients/client-java/src/main/java/org/apache/gravitino/client/GenericModel.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import com.google.common.collect.Lists;
import java.util.List;
import java.util.Map;
import org.apache.gravitino.Audit;
import org.apache.gravitino.MetadataObject;
import org.apache.gravitino.MetadataObjects;
import org.apache.gravitino.Namespace;
import org.apache.gravitino.authorization.SupportsRoles;
import org.apache.gravitino.dto.model.ModelDTO;
import org.apache.gravitino.exceptions.NoSuchPolicyException;
import org.apache.gravitino.exceptions.NoSuchTagException;
import org.apache.gravitino.exceptions.PolicyAlreadyAssociatedException;
import org.apache.gravitino.exceptions.TagAlreadyAssociatedException;
import org.apache.gravitino.model.Model;
import org.apache.gravitino.policy.Policy;
import org.apache.gravitino.policy.SupportsPolicies;
import org.apache.gravitino.tag.SupportsTags;
import org.apache.gravitino.tag.Tag;

/** Represents a generic model. */
class GenericModel implements Model, SupportsTags, SupportsPolicies {

  private final ModelDTO modelDTO;

  private final MetadataObjectTagOperations objectTagOperations;
  private final MetadataObjectPolicyOperations objectPolicyOperations;

  GenericModel(ModelDTO modelDTO, RESTClient restClient, Namespace modelNs) {
    this.modelDTO = modelDTO;
    List<String> modelFullName =
        Lists.newArrayList(modelNs.level(1), modelNs.level(2), modelDTO.name());
    MetadataObject modelObject = MetadataObjects.of(modelFullName, MetadataObject.Type.MODEL);
    this.objectTagOperations =
        new MetadataObjectTagOperations(modelNs.level(0), modelObject, restClient);
    this.objectPolicyOperations =
        new MetadataObjectPolicyOperations(modelNs.level(0), modelObject, restClient);
  }

  @Override
  public Audit auditInfo() {
    return modelDTO.auditInfo();
  }

  @Override
  public String name() {
    return modelDTO.name();
  }

  @Override
  public String comment() {
    return modelDTO.comment();
  }

  @Override
  public Map<String, String> properties() {
    return modelDTO.properties();
  }

  @Override
  public int latestVersion() {
    return modelDTO.latestVersion();
  }

  @Override
  public SupportsTags supportsTags() {
    return this;
  }

  @Override
  public SupportsPolicies supportsPolicies() {
    return this;
  }

  @Override
  public SupportsRoles supportsRoles() {
    throw new UnsupportedOperationException("Not supported yet.");
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof GenericModel)) {
      return false;
    }

    GenericModel that = (GenericModel) o;
    return modelDTO.equals(that.modelDTO);
  }

  @Override
  public int hashCode() {
    return modelDTO.hashCode();
  }

  @Override
  public String[] listTags() {
    return objectTagOperations.listTags();
  }

  @Override
  public Tag[] listTagsInfo() {
    return objectTagOperations.listTagsInfo();
  }

  @Override
  public Tag getTag(String name) throws NoSuchTagException {
    return objectTagOperations.getTag(name);
  }

  @Override
  public String[] associateTags(String[] tagsToAdd, String[] tagsToRemove)
      throws TagAlreadyAssociatedException {
    return objectTagOperations.associateTags(tagsToAdd, tagsToRemove);
  }

  @Override
  public String[] listPolicies() {
    return objectPolicyOperations.listPolicies();
  }

  @Override
  public Policy[] listPolicyInfos() {
    return objectPolicyOperations.listPolicyInfos();
  }

  @Override
  public Policy getPolicy(String name) throws NoSuchPolicyException {
    return objectPolicyOperations.getPolicy(name);
  }

  @Override
  public String[] associatePolicies(String[] policiesToAdd, String[] policiesToRemove)
      throws PolicyAlreadyAssociatedException {
    return objectPolicyOperations.associatePolicies(policiesToAdd, policiesToRemove);
  }
}

File: clients/client-python/gravitino/client/generic_model.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from typing import Optional

from gravitino.api.metadata_object import MetadataObject
from gravitino.api.metadata_objects import MetadataObjects
from gravitino.api.model.model import Model
from gravitino.api.tag import Tag
from gravitino.api.tag.supports_tags import SupportsTags
from gravitino.client.metadata_object_tag_operations import MetadataObjectTagOperations
from gravitino.dto.audit_dto import AuditDTO
from gravitino.dto.model_dto import ModelDTO
from gravitino.namespace import Namespace
from gravitino.utils import HTTPClient


class GenericModel(Model, SupportsTags):
    _model_dto: ModelDTO
    """The model DTO object."""

    def __init__(
        self, model_dto: ModelDTO, rest_client: HTTPClient, model_ns: Namespace
    ) -> None:
        self._model_dto = model_dto
        model_full_name = [model_ns.level(1), model_ns.level(2), model_dto.name()]
        model_object: MetadataObject = MetadataObjects.of(
            model_full_name, MetadataObject.Type.MODEL
        )
        self._model_tag_operations = MetadataObjectTagOperations(
            model_ns.level(0), model_object, rest_client
        )

    def __eq__(self, value: object) -> bool:
        if not isinstance(value, GenericModel):
            return False

        return self._model_dto == value._model_dto

    def __hash__(self) -> int:
        return hash(self._model_dto)

    def name(self) -> str:
        return self._model_dto.name()

    def comment(self) -> Optional[str]:
        return self._model_dto.comment()

    def properties(self) -> dict:
        return self._model_dto.properties()

    def latest_version(self) -> int:
        return self._model_dto.latest_version()

    def audit_info(self) -> AuditDTO:
        return self._model_dto.audit_info()

    def list_tags(self) -> list[str]:
        return self._model_tag_operations.list_tags()

    def list_tags_info(self) -> list[Tag]:
        return self._model_tag_operations.list_tags_info()

    def get_tag(self, name: str) -> Tag:
        return self._model_tag_operations.get_tag(name)

    def associate_tags(
        self, tags_to_add: list[str], tags_to_remove: list[str]
    ) -> list[str]:
        return self._model_tag_operations.associate_tags(tags_to_add, tags_to_remove)
